### ISSUE TITLE
PromiseTask.isCancelled performs an unsynchronized read

### ISSUE THREAD

**Issue by ejona86** [netty/netty#10026]
In #9744 @njhill changed PromiseTask.isCancelled to:

https://github.com/netty/netty/blob/fcf55fcf718af1e4cae85eb86cdb679094186aed/common/src/main/java/io/netty/util/concurrent/PromiseTask.java#L180-L182

But `task` is just a plain field:

https://github.com/netty/netty/blob/fcf55fcf718af1e4cae85eb86cdb679094186aed/common/src/main/java/io/netty/util/concurrent/PromiseTask.java#L65

A race detector is noticing this unsynchronized read/write.

The write is occurring in `clearTaskAfterCompletion()`, which does mention the potential for a _write_ race, which does seem likely to be minor:

https://github.com/netty/netty/blob/fcf55fcf718af1e4cae85eb86cdb679094186aed/common/src/main/java/io/netty/util/concurrent/PromiseTask.java#L114-L123

I've not spent too much time looking at it, but it is unclear to me why `task == CANCELLED` is even checked in `isCancelled()`; it seems like `DefaultPromise.isCancelled()` should be enough. It sort of seems like the semantics of `isCancelled()` are being changed to be more of a `isDone()`. But I don't think I understand the code enough to know how serious this race is.

**Comment by njhill** [comment]
@ejona86 it's just a small optimization to avoid a volatile read in the cancelled case - if `CANCELLED` is seen as the value of `task` then the promise _must_ be cancelled since it is a terminal state. It was motivated by the fact that cancellation is checked more frequently after the #9744 changes (see [here](https://github.com/netty/netty/pull/9744/files#diff-a6871cf1d78dcb4ff1db3d6bb5e117b9R161)), and is expected to be true in more "normal" cases, i.e. whenever a scheduled task is cancelled.

But the fact that it wasn't clear to you means it deserves a comment to be added at minimum! I can happily open a PR for that.

**Comment by ejona86** [comment]
@njhill, okay, that's good to know. I'll look in to seeing if there's a way to whitelist this thread-safety failure.

Do we know if that optimization actually provides any benefit? I see you made benchmarks for the change overall, but I'd expect that optimization to provide ~zero benefit since it only speeds up the `cancelled == true` case and the volatile is still within the normal flow (so the compiler would still be limited in its reordering).

**Comment by ejona86** [comment]
Looks like there is no way to whitelist it, or it will be a headache. (The tools work in Java properly but the tooling people are C++ folks and wouldn't be used to the relatively high safety guaranteed to data races in Java.) If we can show it doesn't benefit much, then we may try to remove the optimization.

**Comment by normanmaurer** [comment]
Let’s just remove it ...

**Comment by ejona86** [comment]
@creamsoup can do the removal, but it will probably be a week or so. If anyone beats us to the punch we wouldn't be bothered :smile:.

**Comment by njhill** [comment]
> it only speeds up the `cancelled == true` case

My point was that case would now be more common. AFAIK timeouts are commonly implemented by scheduling a task for every request then immediately cancelling it when the response is received. And for those cases this may become the "normal" flow.

> Looks like there is no way to whitelist it

I didn't think (careful) use of intentionally-racy logic was uncommon in high-performance java concurrency... so am surprised there aren't other places flagged by the detector. And if other places need to be whitelisted then it seems this may as well be included.

> Let’s just remove it ...

I'm also happy to do that too and will try the benchmark myself in any case!

**Comment by ejona86** [comment]
There used to be more failures in Netty, but many of them have been fixed. I did find a whitelist and it appears the only Netty race whitelisted is from io.netty.util.Recycler$WeakOrderQueue.transfer (not saying such a race still exists). But I do see a reasonable number of races ignored for the java standard library. That might be to just allow the tool to be used; stop new issues from being introduced and clean up preexisting issues over time.

In any case, if we don't get much performance benefit it is probably best to remove the optimization. If we can show the benefit then we can try to get on the whitelist.
